cloud/cloud_manifest.cc

```diff
@@ -175,13 +175,28 @@ Status CloudManifest::WriteToLog(std::unique_ptr<WritableFileWriter> log) {
   return writer.file()->Sync(true);
 }
 
-void CloudManifest::AddEpoch(uint64_t startFileNumber, std::string epochId) {
+bool CloudManifest::AddEpoch(uint64_t startFileNumber, std::string epochId) {
   WriteLock lck(&mutex_);
-  assert(pastEpochs_.empty() || pastEpochs_.back().first <= startFileNumber);
-  if (pastEpochs_.empty() || pastEpochs_.back().first < startFileNumber) {
-      pastEpochs_.emplace_back(startFileNumber, std::move(currentEpoch_));
-  }  // Else current epoch hasn't written any files, I can just ignore it
+  if (!pastEpochs_.empty() && startFileNumber < pastEpochs_.back().first) {
+    return false;
+  }
+
+  // Check all the epochs with same file number
+  // For each (filenum, epoch) pair in `pastEpochs_`, it means next epoch of `epoch`
+  // starts at `filenum`. So we should compare epochId with next epoch of `epoch`.
+  auto nxtEpoch = &currentEpoch_;
+  for (auto rit = pastEpochs_.rbegin(); rit != pastEpochs_.rend(); rit++) {
+    if (rit->first == startFileNumber && *nxtEpoch == epochId) {
+      return false;
+    }
+    if (rit->first < startFileNumber) {
+      break;
+    }
+    nxtEpoch = &(rit->second);
+  }
+  pastEpochs_.emplace_back(startFileNumber, std::move(currentEpoch_));
```

Comment:
Is it okay for us to have two entries in `pastEpochs_` with the same `startFileNumber`?

Reply:
Or actually I see this comment below
```
  // Idempotency is based on the assumption that epochs added don't diverge from
  // existing epochs(same sequence of <filenumm, epoch> is re-added)
```
So for every call to `AddEpoch()` with a given `startFileNumber` and `epochId`, we assume that all calls to `AddEpoch()` with same `epochId` will also have the same `startFileNumber`?

Reply:
It's possible to have two entries in `pastEpochs` with same file number but different epoch. For example, a db is closed and then reopened without generating any new file. The newly added two tests try to cover this case for both LF and non LF mode.